ebpf/python/pyperf.go

```diff
@@ -0,0 +1,279 @@
+package python
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/perf"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/pyroscope/ebpf/metrics"
+	"github.com/grafana/pyroscope/ebpf/symtab"
+	lru "github.com/hashicorp/golang-lru/v2"
+)
+
+type Perf struct {
+	rd             *perf.Reader
+	logger         log.Logger
+	pidDataHashMap *ebpf.Map
+	symbolsHashMp  *ebpf.Map
+	metrics        *metrics.PythonMetrics
+
+	events      []*PerfPyEvent
+	eventsLock  sync.Mutex
+	sc          *symtab.SymbolCache
+	pidCache    *lru.Cache[uint32, *PerfPyPidData]
+	prevSymbols map[uint32]*PerfPySymbol
+	wg          sync.WaitGroup
+}
+
+func NewPerf(logger log.Logger, metrics *metrics.PythonMetrics, perfEventMap *ebpf.Map, pidDataHasMap *ebpf.Map, symbolsHashMap *ebpf.Map) (*Perf, error) {
+	rd, err := perf.NewReader(perfEventMap, 4*os.Getpagesize())
+	if err != nil {
+		return nil, fmt.Errorf("perf new reader: %w", err)
+	}
+	pidCache, err := lru.NewWithEvict[uint32, *PerfPyPidData](512, func(key uint32, value *PerfPyPidData) {
+		_ = pidDataHasMap.Delete(key)
+	})
+	if err != nil {
+		return nil, fmt.Errorf("pyperf pid cache %w", err)
+	}
+	res := &Perf{
+		rd:             rd,
+		logger:         logger,
+		pidDataHashMap: pidDataHasMap,
+		symbolsHashMp:  symbolsHashMap,
+		pidCache:       pidCache,
+		metrics:        metrics,
+	}
+	res.wg.Add(1)
+	go func() {
+		defer res.wg.Done()
+		res.loop()
+	}()
+	return res, nil
+}
+
+// AddPythonPID return python init error, second return value is true if error occured and process is stil alive,
+// third return value - python process data
+func (s *Perf) AddPythonPID(pid uint32, serviceName string) (error, bool, *PerfPyPidData) {
```

Comment:
err should be the last returned element not the first. It's also not clear where you add the PID.